include/mfmg/instantiation.hpp

```diff
@@ -0,0 +1,44 @@
+/*************************************************************************
+ * Copyright (c) 2017 by the mfmg authors                                *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * This file is part of the mfmg libary. mfmg is distributed under a BSD *
+ * 3-clause license. For the licensing terms see the LICENSE file in the *
+ * top-level directory                                                   *
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#ifndef AMGE_INSTANTIATION_HPP
+#define AMGE_INSTANTIATION_HPP
+
+#include <deal.II/lac/la_parallel_vector.h>
+#include <deal.II/lac/trilinos_vector.h>
+
+#include <boost/preprocessor/repeat_from_to.hpp>
+#include <boost/preprocessor/seq/for_each.hpp>
+#include <boost/preprocessor/tuple/replace.hpp>
+
+// TUPLE(class_name) expend to (class_name, 0)
+#define TUPLE_N (0, 0)
```

Comment:
OK, I don't understand any of the code in this file. We'll have to find some time to sit down for a chat.

Reply:
OK :-) the big thing is that you need to read this bottom up. That's how the loops are executed.